cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -148,128 +191,197 @@ struct ListSlice {
-        RETURN_NOT_OK(BuildArray<LargeListBuilder>(batch, opts, *builder));
-      } else {
-        RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), list(value_type), &builder));
-        RETURN_NOT_OK(BuildArray<ListBuilder>(batch, opts, *builder));
-      }
+    auto* pool = ctx->memory_pool();
+    ARROW_ASSIGN_OR_RAISE(auto output_type_holder, ListSliceOutputType(opts, *list_type));
+    constexpr auto kInputTypeId = InListType::type_id;
+    auto output_type = output_type_holder.GetSharedPtr();
+    switch (output_type->id()) {
+      case Type::LIST:
+        DCHECK(kInputTypeId == Type::LIST || kInputTypeId == Type::FIXED_SIZE_LIST);
+        if constexpr (kInputTypeId == Type::LIST ||
+                      kInputTypeId == Type::FIXED_SIZE_LIST) {
+          return BuildArray<ListBuilder>(pool, opts, batch, output_type, out);
+        }
+        break;
+      case Type::LARGE_LIST:
+        DCHECK_EQ(kInputTypeId, Type::LARGE_LIST);
+        if constexpr (kInputTypeId == Type::LARGE_LIST) {
+          return BuildArray<LargeListBuilder>(pool, opts, batch, output_type, out);
+        }
+        break;
+      case Type::FIXED_SIZE_LIST:
+        // A fixed-size list can be produced from any list-like input
+        // if ListSliceOptions::return_fixed_size_list is set to true
+        return BuildArray<FixedSizeListBuilder>(pool, opts, batch, output_type, out);
+      case Type::LIST_VIEW:
+        DCHECK_EQ(kInputTypeId, Type::LIST_VIEW);
+        if constexpr (kInputTypeId == Type::LIST_VIEW) {
+          return BuildArray<ListViewBuilder>(pool, opts, batch, output_type, out);
+        }
+        break;
+      case Type::LARGE_LIST_VIEW:
+        DCHECK_EQ(kInputTypeId, Type::LARGE_LIST_VIEW);
+        if constexpr (kInputTypeId == Type::LARGE_LIST_VIEW) {
+          return BuildArray<LargeListViewBuilder>(pool, opts, batch, output_type, out);
+        }
+        break;
+      default:
+        break;
     }
-
-    // build output arrays and set result
-    ARROW_ASSIGN_OR_RAISE(auto result, builder->Finish());
-    out->value = std::move(result->data());
+    Unreachable();
     return Status::OK();
   }
 
+  /// \brief Builds the array of list slices from the input list array
   template <typename BuilderType>
-  static Status BuildArray(const ExecSpan& batch, const ListSliceOptions& opts,
-                           ArrayBuilder& builder) {
-    if constexpr (std::is_same_v<Type, FixedSizeListType>) {
-      RETURN_NOT_OK(BuildArrayFromFixedSizeListType<BuilderType>(batch, opts, builder));
+  static Status BuildArray(MemoryPool* pool, const ListSliceOptions& opts,
+                           const ExecSpan& batch,
+                           const std::shared_ptr<DataType>& output_type,
+                           ExecResult* out) {
+    std::unique_ptr<ArrayBuilder> builder;
+    RETURN_NOT_OK(MakeBuilder(pool, output_type, &builder));
+    auto* list_builder = checked_cast<BuilderType*>(builder.get());
```

Comment:
Should we start by presizing the builder?

Reply:
Can't believe I've forgot this one. It's 🚀 on the benchmarks now.